docs: Update LayoutLM model card according to new standardized format (#40129)
huggingface/transformers: docs/source/en/model_doc/layoutlm.md

It's also pre-trained using two unique approaches:

- **Masked Visual-Language Model (MVLM)**: It's like a fill-in-the-blanks test. The model sees a document with some words blacked out, and it has to guess what's missing by looking at the other words and their positions on the page.

- **Multi-label Document Classification (MDC)**: It also learns to categorize entire documents (like "this is a tax form" or "this is a resume"). This helps it get a better overall understanding of the document's structure and purpose.

So, in a nutshell, it's a model that understands both the words and the layout, making it a great tool for reading and understanding documents.

You can find all the original LayoutLM checkpoints under the [LayoutLM](https://huggingface.co/collections/microsoft/layoutlm-6564539601de72cb631d0902) collection.

> [!TIP]

> > This model was contributed by [microsoft](https://github.com/microsoft).
> >

Reviewer: ```suggestion
```

> Click on the LayoutLM models in the right sidebar for more examples of how to apply LayoutLM to different vision and language tasks.

> The example below demonstrates how to use the model with the ([`Pipeline`](https://huggingface.co/docs/transformers/main/en/main_classes/pipelines#transformers.Pipeline) or) [`AutoModel`](https://huggingface.co/docs/transformers/main/en/model_doc/auto#transformers.AutoModel) class. 

Reviewer: ```suggestion
The example below demonstrates question answering with the [`AutoModel`] class. 
```

<hfoptions id="usage">
<hfoption id="Pipeline">

Note, that the original LayoutLM version cannot be readily used with the [`Pipeline`] class unlike later versions such as [**LayoutLMv3**](https://huggingface.co/docs/transformers/en/model_doc/layoutlmv3). 

</hfoption>
<hfoption id="AutoModel">

> ```py

Reviewer: More helpful to show a document answering example

```py
import torch
from datasets import load_dataset
from transformers import AutoTokenizer, LayoutLMForQuestionAnswering

tokenizer = AutoTokenizer.from_pretrained("impira/layoutlm-document-qa", add_prefix_space=True)
model = LayoutLMForQuestionAnswering.from_pretrained("impira/layoutlm-document-qa", torch_dtype=torch.float16)

dataset = load_dataset("nielsr/funsd", split="train")
example = dataset[0]
question = "what's his name?"
words = example["words"]
boxes = example["bboxes"]

encoding = tokenizer(
    question.split(), words, is_split_into_words=True, return_token_type_ids=True, return_tensors="pt"
)
bbox = []
for i, s, w in zip(encoding.input_ids[0], encoding.sequence_ids(0), encoding.word_ids(0)):
    if s == 1:
        bbox.append(boxes[w])
    elif i == tokenizer.sep_token_id:
        bbox.append([1000] * 4)
    else:
        bbox.append([0] * 4)
encoding["bbox"] = torch.tensor([bbox])

word_ids = encoding.word_ids(0)
outputs = model(**encoding)
loss = outputs.loss
start_scores = outputs.start_logits
end_scores = outputs.end_logits
start, end = word_ids[start_scores.argmax(-1)], word_ids[end_scores.argmax(-1)]
print(" ".join(words[start : end + 1]))
```

from transformers import AutoTokenizer, AutoModelForTokenClassification
from datasets import load_dataset

# Load a sample dataset to get an example
dataset = load_dataset("nielsr/funsd")
example = dataset["train"][0]
words = example["words"] # provides already pre-processed inputs for demonstration purposes
boxes = example["bboxes"] # provides already pre-processed inputs for demonstration purposes

# Load model and tokenizer
tokenizer = AutoTokenizer.from_pretrained("microsoft/layoutlm-base-uncased", use_fast=False)
model = AutoModelForTokenClassification.from_pretrained("microsoft/layoutlm-base-uncased")

# Process inputs
encoding = tokenizer(words, boxes=boxes, is_split_into_words=True, return_tensors="pt")

# Get predictions
outputs = model(**encoding)

print(outputs.logits.shape) # (1, 148, 7) - batch_size, sequence_length, num_labels
```

</hfoption>

> <hfoption id="transformers-cli">

Reviewer: This can be removed as well

Author: I don't know why the part about bounding normalization got here..., but I guess you mean that I can delete the part about transformer-cli, right? :)

Reviewer: Yeah I was referring to the transformers CLI part!

For the original LayoutLM version, `transformers-cli` usage is not readily applicable unlike later versions such as [**LayoutLMv3**](https://huggingface.co/docs/transformers/en/model_doc/layoutlmv3). 

</hfoption>
</hfoptions>

> Quantization reduces the memory burden of large models by representing the weights in a lower precision. 

Reviewer: The model isn't that large so no need to include a quantization example

The example below uses `TorchAoConfig` to only quantize the weights to `int8`.

```python
# pip install torchao accelerate
import torch
from transformers import TorchAoConfig, LayoutLMForTokenClassification, AutoTokenizer
from datasets import load_dataset

# Define the quantization configuration
quantization_config = TorchAoConfig("int8_weight_only", group_size=128)

# Load the model with 8-bit quantization
model = LayoutLMForTokenClassification.from_pretrained(
    "microsoft/layoutlm-base-uncased",
    torch_dtype=torch.bfloat16,
    quantization_config=quantization_config
)

# Load the tokenizer
tokenizer = AutoTokenizer.from_pretrained("microsoft/layoutlm-base-uncased", use_fast=False)

# Example input
dataset = load_dataset("nielsr/funsd")
example = dataset["train"][0]
encoding = tokenizer(example["words"], boxes=example["bboxes"], is_split_into_words=True, return_tensors="pt")
inputs = encoding

# Perform inference
outputs = model(**inputs)
print(outputs)
```

## Notes

> - The original LayoutLM was not designed with a unified processing workflow. Instead, it expects you to pre-process text (`words`) and bounding boxes (`boxes`) via an external OCR engine, e.g. Google's Tesseract (there’s a [Python wrapper](https://pypi.org/project/pytesseract/) available) and provide them as additional inputs to the tokenizer. 

Reviewer: ```suggestion
- The original LayoutLM was not designed with a unified processing workflow. Instead, it expects preprocessed text (`words`) and bounding boxes (`boxes`) from an external OCR engine (like [Pytesseract](https://pypi.org/project/pytesseract/)) and provide them as additional inputs to the tokenizer. 
```

> For simplified demonstration purposes, the above code examples use example data from the `datasets` library, which come along with already pre-processed words and boxes. 
> 
> - For more details and examples of usage refer to the links in the [Resources](#resources) section. 

Reviewer: ```suggestion
```

> - **Bounding Boxes**: Bounding box coordinates should be in the format `(x_min, y_min, x_max, y_max)`. It's best practice to normalize these coordinates to a range like `[0, 1000]`. To normalize, you can use the following function:

Reviewer: ```suggestion
- The [`~LayoutLM.forward`] method expects the input `bbox` (bounding boxes of the input tokens). Each bounding box should be in the format `(x0, y0, x1, y1)`.  `(x0, y0)` corresponds to the upper left corner of the bounding box and `{x1, y1)` corresponds to the lower right corner. The bounding boxes need to be normalized on a 0-1000 scale as shown below.
```

    ]
```

> - `width` and `height` correspond to the width and height of the original document in which the token occurs. Those can be obtained using the Python Image Library (PIL) library for example, as follows:

Reviewer: ```suggestion
- `width` and `height` correspond to the width and height of the original document in which the token occurs. These values can be obtained as shown below.
```

width, height = image.size
```

> ## Resources

Reviewer: For the Resources section, let's keep the external ones. You can just use a simple list for them:

```md
- Read [fine-tuning LayoutLM for document-understanding using Keras & Hugging Face Transformers](https://www.philschmid.de/fine-tuning-layoutlm-keras) to learn more.
- Read [fine-tune LayoutLM for document-understanding using only Hugging Face Transformers](https://www.philschmid.de/fine-tuning-layoutlm) for more information.
- Refer to this [notebook](https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Add_image_embeddings_to_LayoutLM.ipynb) for a practical example of how to fine-tune LayoutLM.
- Refer to this [notebook](https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Fine_tuning_LayoutLMForSequenceClassification_on_RVL_CDIP.ipynb) for an example of how to fine-tune LayoutLM for sequence classification.
- Refer to this [notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Fine_tuning_LayoutLMForTokenClassification_on_FUNSD.ipynb) for an example of how to fine-tune LayoutLM for token classification.
- Read [Deploy LayoutLM with Hugging Face Inference Endpoints](https://www.philschmid.de/inference-endpoints-layoutlm) to learn how to deploy LayoutLM.
```

Author: Ok, done :)

> - **Tokenization**: The model uses the standard `BertTokenizer` to tokenize the text (, and you'll need to pass the bounding box information for each token). 

Reviewer: ```suggestion
```

> - [**AttentionMaskVisualizer**](https://github.com/huggingface/transformers/blob/beb9b5b02246b9b7ee81ddf938f93f44cfeaad19/src/transformers/utils/attention_visualizer.py#L139) is not directly compatible with LayoutLM's multi-modal output (text and layout) and attention structure (as opposed to text-based models like BERT with standard attention mechanisms).  

Reviewer: Remove the `AttentionMaskVisualizer` section since its unsupported

Author: Thanks, done.

> - The AutoProcessor in [**LayoutLMv3**](https://huggingface.co/docs/transformers/en/model_doc/layoutlmv3) greatly simplifies the workflow by integrating all mentioned steps into a single, user-friendly class. 

Reviewer: ```suggestion
```

> ### LayoutLMConfig

Reviewer: Don't need to change any of the API section

Author: Great, thanks! :)
